geomesa-fs/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/PartitionScheme.scala

```diff
@@ -0,0 +1,332 @@
+
+      case "hourly" =>
+        new DateTimeScheme(Hourly, ChronoUnit.HOURS, 1, sft.getDtgField.get, true)
+
+      case "daily" =>
+        new DateTimeScheme(Daily, ChronoUnit.DAYS, 1, sft.getDtgField.get, true)
+
+      case "monthly" =>
+        new DateTimeScheme(Monthly, ChronoUnit.MONTHS, 1, sft.getDtgField.get, true)
+
+      case z2 if z2.matches("z2-[0-9]+bit") =>
+        val bits = "z2-([0-9]+)bit".r("bits").findFirstMatchIn(z2).get.group("bits").toInt
+        new Z2Scheme(bits, sft.getGeomField, true)
+
+      case _ =>
+        throw new IllegalArgumentException(s"Unable to find well known scheme(s) for argument $name")
+    }
+    if (schemes.length == 1) {
+      schemes.head
+    } else {
+      new CompositeScheme(schemes.toSeq)
+    }
+  }
+}
+
+object PartitionScheme {
+
+  val SchemeSeparator = ","
+
+  // Must begin with GeoMesa in order to be persisted
+  val PartitionSchemeKey = "geomesa.fs.partition-scheme.config"
+  val PartitionOptsPrefix = "fs.partition-scheme.opts."
+  val PartitionSchemeParam = new Param("fs.partition-scheme.name", classOf[String], "Partition scheme name", false)
+
+  def addToSft(sft: SimpleFeatureType, scheme: PartitionScheme): Unit =
+    sft.getUserData.put(PartitionSchemeKey, scheme.toString)
+
+  def extractFromSft(sft: SimpleFeatureType): PartitionScheme = {
+    if (!sft.getUserData.containsKey(PartitionSchemeKey)) {
+      throw new IllegalArgumentException("SFT does not have partition scheme in hints")
+    }
+    apply(sft, sft.getUserData.get(PartitionSchemeKey).asInstanceOf[String])
+  }
+
+  def apply(sft: SimpleFeatureType, dsParams: util.Map[String, Serializable]): PartitionScheme = {
+    val pName = PartitionSchemeParam.lookUp(dsParams).toString
+    import scala.collection.JavaConversions._
+    val pOpts = dsParams.keySet.filter(_.startsWith(PartitionOptsPrefix)).map { opt =>
+      opt.replace(PartitionOptsPrefix, "") -> dsParams.get(opt).toString
+    }.toMap
+    PartitionScheme(sft, pName, pOpts)
+  }
+
+  // TODO delegate out, etc. make a loader, etc
+  def apply(sft: SimpleFeatureType, pName: String, opts: Map[String, String]): PartitionScheme = {
+    import PartitionOpts._
+    val leaf = parseLeafStorage(opts)
+    val schemes = pName.split(SchemeSeparator).map {
+      case DateTimeScheme.Name =>
+        val attr = parseDtgAttr(opts)
+        val fmt = parseDateTimeFormat(opts)
+        val su = parseStepUnit(opts)
+        val s = parseStep(opts)
+        new DateTimeScheme(fmt, su, s, attr, leaf)
+
+      case Z2Scheme.Name =>
+        val geomAttr = parseGeomAttr(opts)
+        val z2Res = parseZ2Resolution(opts)
+        new Z2Scheme(z2Res, geomAttr, leaf)
+
+      case _ =>
+        throw new IllegalArgumentException(s"Unknown scheme name $pName")
+    }
+
+    if (schemes.length == 1) {
+      schemes.head
+    } else {
+      new CompositeScheme(schemes.toSeq)
+    }
+  }
+
+  // TODO meh this sucks
```

Comment:
Alright....  fix/address the todo or make it more descriptive so that someone could;).